flasc/data_processing/northing_offset_change_hoger.py

```diff
@@ -0,0 +1,281 @@
+
+    # Reference date
+    if reference == "first":
+        ref = 0
+    elif reference == "last":
+        ref = df.shape[0] - 1
+    else:
+        date_ref = pd.to_datetime(reference)
+        ref = df["time"][df["time"] == date_ref]
+        if len(ref) == 0:
+            if date_ref < df["time"].min():
+                ref = 0
+            elif date_ref > df["time"].max():
+                ref = df.shape[0] - 1
+            else:
+                ref = 0
+                warnings.warn(
+                    "The reference date seem to be missing in the dataset. "
+                    " The first date is selected as reference."
+                )
+
+    # Build the DataFrame to store the results of the Tree
+    d = pd.DataFrame(columns=["Knot", "Jump", "Turbine"])
+    d = d.astype({"Knot": np.float64, "Jump": np.float64, "Turbine": str})  # Assign types
+
+    # Iterate over every wind turbine comparing its direction with every other.
+    #  Get then the points at which
+    # a jump is produced. These points are called knots.
+    for m in wt_names:
+        # Get the wind turbines to compare
+        ms2 = wt_names[wt_names != m]
+        df2 = df[ms2]
+        for m2 in ms2:
+            # Get the differences in the direction
+            df2.loc[:, m2] = wrap_180(df[m2] - df[m])
+            y = df2[m2][~df2[m2].isna()]  # Do not use the nan values
+            # Use a decision tree regressor to get the points at which there are knots
+            # and the values of the
+            # direction differences
+            regr = DecisionTreeRegressor(
+                max_depth=max_depth,
+                min_samples_split=threshold,
+                min_samples_leaf=threshold // 2,
+                ccp_alpha=ccp_alpha,
+            )
+            regr.fit(y.index.to_numpy()[:, np.newaxis], y)
+
+            # Postprocess the results of the decision tree
+            # regressor to get the information required.
+            # The leaves are
+            # the values of the means and the knots the points at which jumps occur
+            leaves, knots = _get_leaves_and_knots(regr)
+
+            # Plot the results if desired
+            if plot_it:
+                _plot_regression(
+                    df2[m2],
+                    regr.predict(y.index.to_numpy()[:, np.newaxis]),
+                    scada.time,
+                    f"{m2} - {m} wind direction [°]",
+                )
+
+            # Save the results
+            if len(knots) > 0 and not np.any(np.isnan(knots)):
+                if m == "wd_004":
+                    print(knots)
+                d = pd.concat(
+                    [d, pd.DataFrame({"Knot": knots, "Jump": np.diff(leaves), "Turbine": m})]
+                )
+
+    d = d.reset_index(drop=True)
+
+    # Postprocess all the data to get the main jumps for each wind turbine
+    d2 = d.copy()
+    d2["Class"] = discretize(d["Knot"], threshold=100)
```

Comment:
The value of the threshold argument for the discretize function should be the same as the one used for the decission tree regressor. It was a mistake on my part. Thus,

`d2["Class"] = discretize(d["Knot"], threshold=threshold)`

would be more appropiate.

Reply:
Thanks for catching, fixed!